### PR TITLE
fix: timeout stuck upload on bytes endpoint

### LINKED ISSUE #2946: Direct upload gets stuck forever
### Context
Uploading a file with the `swarm-deferred-upload` header set to `false` (also known as direct upload) may get stuck forever depending on the data to be uploaded. Uploading the same data with `swarm-deferred-upload` set to `true` (which is the default) succeeds and it seems to be cached after, because once it was successfully uploaded then the following uploads are going to be successful no matter how the header is set. 

### Steps to reproduce
There is an attached zip file which contains a file, which has 16384 bytes size and has `4d36fd00cd3d1eb376c2633a10a27287c90f48e696e401f8dec23548031a6f16` as content-hash.

Use the contained file and upload it to the `/bytes` endpoint with the `swarm-deferred-upload` header set to false:
```
curl --data-binary @4d36fd00cd3d1eb376c2633a10a27287c90f48e696e401f8dec23548031a6f16.bin -H "swarm-deferred-upload: false" -H "swarm-postage-batch-id: <use-your-postage-batch-id>" http://localhost:1633/bytes
```
[4d36fd00cd3d1eb376c2633a10a27287c90f48e696e401f8dec23548031a6f16.zip](https://github.com/ethersphere/bee/files/8662972/4d36fd00cd3d1eb376c2633a10a27287c90f48e696e401f8dec23548031a6f16.zip)

### PR DESCRIPTION
**Pull request by mrekucci**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [x] I have added tests to cover my changes.

### Related Issue (Optional)
Closes #2946

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3041)

### CODE REVIEW

**Review by notanatol — APPROVED** [review]

File: pkg/api/util.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+package api
+
+import (
+	"context"
+	"io"
+	"time"
+)
+
+// ProgressReader tracks the progress on io.Reader Read method.
+type ProgressReader struct {
```
Comment (janos) [inline pkg/api/util.go:14]: I would suggest not to export types and functions unless that is required. Currently, ProgressReader and its constructor is only used in the api package. Maybe there is even a better place for ProgressReader than api package, as it is more general type, not specific to the api package.

**Review by janos — APPROVED** [review]

**Review by acud — APPROVED** [review]
LGTM minus the comment from @janos

File: pkg/api/util.go (hunk 1 above)
Comment (mrekucci) [reply pkg/api/util.go:14]: Agree, I exported the types and functions because of the `api_test` package. Moved to its own general package.

**Review by aloknerurkar — APPROVED** [review]
Nicely done!

**Review by istae — APPROVED** [review]

**Pull request merged**